autorest/azure/async.go

```diff
@@ -0,0 +1,165 @@
+	// Properties, on operation success, optionally contains per-service / per-resource values.
+	Properties map[string]interface{} `json:"properties,omitempty"`
+	// Error provides additional detail if the operation is canceled or failed.
+	Error OperationError `json:"error,omitempty"`
+	// StartTime optionally provides the time the operation started.
+	StartTime date.Time `json:"startTime,omitempty"`
+	// EndTime optionally provides the time the operation completed.
+	EndTime date.Time `json:"endTime,omitempty"`
+	// PercentComplete optionally provides the percent complete between 0 and 100.
+	PercentComplete float64 `json:"percentComplete,omitempty"`
+}
+
+// IsCanceled returns true if the operation was canceled; false otherwise.
+func (or OperationResource) IsCanceled() bool {
+	return or.Status == OperationCanceled
+}
+
+// HasFailed returns true if the operation has failed; false otherwise.
+func (or OperationResource) HasFailed() bool {
+	return or.Status == OperationFailed
+}
+
+// HasSucceeded returns true if the operation has succeeded; false otherwise.
+func (or OperationResource) HasSucceeded() bool {
+	return or.Status == OperationSucceeded
+}
+
+// HasTerminated returns true if the operation has terminated; false otherwise.
+func (or OperationResource) HasTerminated() bool {
+	return or.IsCanceled() || or.HasFailed() || or.HasSucceeded()
+}
+
+// GetError returns an error if the operation was canceled or failed and nil otherwise.
+func (or OperationResource) GetError() error {
+	switch {
+	case or.IsCanceled():
+		return fmt.Errorf("azure: Operation was canceled -- error(%v)", or.Error.ToError())
+	case or.HasFailed():
+		return fmt.Errorf("azure: Operation failed -- error(%v)", or.Error.ToError())
+	}
+	return nil
+}
+
+// GetAsyncOperation retrieves the long-running URL from which to retrieve the OperationResource.
+func GetAsyncOperation(resp *http.Response) string {
+	return resp.Header.Get(http.CanonicalHeaderKey(HeaderAsyncOperation))
+}
+
+// IsAsynchronousResponse returns true if the passed response indicates that the request will
+// complete asynchronously. Such responses have either an http.StatusCreated or an
+// http.StatusAccepted status code and provide the Azure-AsyncOperation header.
+func IsAsynchronousResponse(resp *http.Response) bool {
+	return autorest.ResponseHasStatusCode(resp, http.StatusCreated, http.StatusAccepted) &&
+		GetAsyncOperation(resp) != ""
+}
+
+// NewOperationResourceRequest allocates and returns a new http.Request to retrieve the
+// OperationResource for an asynchronous operation.
+func NewOperationResourceRequest(resp *http.Response, cancel <-chan struct{}) (*http.Request, error) {
+	location := GetAsyncOperation(resp)
+	if location == "" {
+		return nil, autorest.NewErrorWithResponse("azure", "NewOperationResourceRequest", resp, "Azure-AsyncOperation header missing from response that requires polling")
+	}
+
+	req, err := autorest.Prepare(&http.Request{Cancel: cancel},
```

Comment:
so this `req, err := ...` statement repeats with `NewPollingRequest`, perhaps extract this into an unexported method?